core/cli/src/rc-file.ts

```diff
@@ -25,27 +26,54 @@ type RawRCFile = {
     | null
 }
 
-export async function loadToolKitRC(logger: Logger, root: string, isAppRoot: boolean): Promise<RCFile> {
-  const result = await explorer.search(root)
+// yaml will automatically stringify any symbols in keys so just use strings
+// that won't be used normally
+export const toolKitOptionIdent = '__toolkit/option__'
+export const toolKitDefinedIdent = '__toolkit/defined__'
 
-  if (!result?.config) {
-    return emptyConfig
-  }
-  if (isAppRoot) {
-    rootConfig = result.filepath
-  } else if (result.filepath === rootConfig) {
-    // Make sure that custom plugins which don't have a config file won't cause
-    // the resolver to use the root config instead and start an infinite loop
-    // of config resolution.
-    logger.warn(
-      `plugin at ${s.filepath(path.dirname(root))} has no config file. please add an empty ${s.filepath(
-        '.toolkitrc'
-      )} file to avoid potential config resolution issues.`
-    )
-    return emptyConfig
+// minimally define the two custom tags so that yaml will parse them without
+// warning but will never be use them when stringifying
+const toolKitOption = {
+  identify: () => false,
+  tag: '!toolkit/option',
+  resolve: (value) => value
+} satisfies YAML.ScalarTag
+const toolKitDefined = {
+  identify: () => false,
+  tag: '!toolkit/defined',
+  resolve: (value) => value
+} satisfies YAML.ScalarTag
+
+export async function loadToolKitRC(logger: Logger, root: string): Promise<RCFile> {
+  let rawConfig: string
+  try {
+    rawConfig = await fs.readFile(path.join(root, '.toolkitrc.yml'), 'utf8')
+  } catch (err) {
+    if ((err as NodeJS.ErrnoException).code === 'ENOENT') {
+      return emptyConfig
+    } else {
+      throw err
+    }
   }
 
-  const config: RawRCFile = result.config
+  const configDoc = YAML.parseDocument(rawConfig, { customTags: [toolKitOption, toolKitDefined] })
+  // go back and search for the parsed tags and replace them with our
+  // indentifiers so we can resolve them once we've loaded plugin options
+  YAML.visit(configDoc, {
+    Scalar(_, node) {
+      if (node.tag === '!toolkit/option') {
+        return configDoc.createNode({ [toolKitOptionIdent]: node.value })
```

Comment:
**question**: why is this done in a `YAML.visit` and not in the `resolve` method of the tags? i did a quick test of that and it preserves symbols as well:

![Screenshot 2024-04-22 at 15 07 23](https://github.com/Financial-Times/dotcom-tool-kit/assets/631757/39b66ea9-9efd-4fb5-9e52-a463554b64b5)

Reply:
**suggestion**: actually, instead of using objects with a specific key here, we could use classes with a property for the option key and the body of the config to substitute. that would also avoid having to concatenate the option key for `toolkit/defined`. doing this would make `deeplySubstitute` a lot cleaner, you could do things like this instead of having to sniff around the object entries:

```js
if(node instanceof OptionsPlaceholder) {
  return resolveOptionPath(node.key)
}
```

Reply:
this might work for `!toolkit/option`, but I don't think it will for `!toolkit/is-defined` 😭 `is-defined` is currently tagging the key of a key-value pair within an object (or a 'map' in YAML nomenclature) and whatever is returned from its `resolve` callback is placed back in the key's position and subsequently stringified (along with a warning) when converted to a JS object. so we can't stuff an object in like we can for `!toolkit/option` and even a symbol is stringified, making these identifiers the least-worst solution. there may be a work around that limitation by finding another part of the YAML schema to tag with `!toolkit/is-defined`, but during my investigation i couldn't find anywhere else that would result in a readable, declarative syntax like we have now.